src/core/qgsconditionalstyle.cpp

```diff
@@ -23,7 +23,37 @@
 
 QgsConditionalLayerStyles::QgsConditionalLayerStyles( QObject *parent )
   : QObject( parent )
-{}
+{
+}
+
+QgsConditionalStyle QgsConditionalLayerStyles::constraintFailureStyles( QgsFieldConstraints::ConstraintStrength strength )
+{
+  switch ( strength )
+  {
+    case QgsFieldConstraints::ConstraintStrengthHard:
+      if ( !mHardConstraintFailureStyle )
+      {
+        mHardConstraintFailureStyle = std::make_unique<QgsConditionalStyle>();
```

Comment:
is the member to avoid having to create the style with every call? If so, I doubt it will make any noticable difference since we're returning a copy here anyway, so we'll incur the cost of copying the style which should be comparable to constructing a new one...

(also, I guess ideally these styles would vary for dark/light themes, but we can tackle that if it's an issue later)